File: Communicators_solution/communicators_solution.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
  int rank, size;
  MPI_Comm comm, new_comm;
  int colour, key;
  int sum_ranks, new_rank;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

/* even ranked processes will supply a colour of 0 and odd ranked, 1 */
  colour = rank%2; 

/* we have no need to fix the ranks in the new communicator */
  key = 0; 

/* the communicator we are splitting */
  comm = MPI_COMM_WORLD; 

/* split the communicator */
  MPI_Comm_split(comm, colour, key, &new_comm);

/* find the rank of the process in the new communicator */
  MPI_Comm_rank(new_comm, &new_rank);

/* print the ranks of each process in the original and new communicators */
  printf("original rank = %d; new rank = %d\n", rank, new_rank);

/* note that the result is being put on rank 0 in the new communicator */
  MPI_Reduce(&rank, &sum_ranks, 1, MPI_INT, MPI_SUM, 0, new_comm);

/* rank 0 in both the odd and even new communicators prints out the result */
  if (new_rank == 0)
    printf("Sum of ranks=%d\n", sum_ranks);

  MPI_Finalize();
}

File: Communicators_solution/communicators_solution.f90
PROGRAM COMMUNICATORS

  USE MPI
  IMPLICIT NONE

  INTEGER :: rank, size, ierror
  INTEGER :: comm, new_comm
  INTEGER :: colour, key
  INTEGER :: sum_ranks, newrank

  CALL MPI_INIT(ierror)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierror)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, size, ierror)

! even ranked processes will supply a colour of 0 and odd ranked, 1
  colour = MOD(rank,2)

! we have no need to fix the ranks in the new communicator
  key = 0

! the communicator we are splitting
  comm = MPI_COMM_WORLD

! split the communicator
  CALL MPI_COMM_SPLIT(comm, colour, key, new_comm, ierror)

! find the rank of the process in the new communicator
  CALL MPI_COMM_RANK(new_comm, newrank, ierror)

! print the ranks of each process in the original and new communicators
  WRITE(*,*) 'original rank = ',rank,'; new rank = ',newrank

! note that the result is being put on rank 0 in the new communicator
  CALL MPI_REDUCE(rank, sum_ranks, 1, MPI_INTEGER, MPI_SUM, 0, new_comm, ierror)

! rank 0 in both the odd and even new communicators prints out the result
  IF (newrank == 0) THEN
    WRITE(*,*) 'Sum of ranks=', sum_ranks
  ENDIF

  CALL MPI_FINALIZE(ierror)

END PROGRAM

